### PR TITLE
feat(rpc): implement Filecoin.EthGetCode

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

Changes introduced in this pull request:

- implement RPC method `Filecoin.EthGetCode`
- API compare tests

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/rpc/methods/eth.rs [hunk 1]
```diff
@@ -1373,6 +1283,73 @@ impl RpcMethod<0> for EthSyncing {
...
+    async fn handle(
+        ctx: Ctx<impl Blockstore + Send + Sync + 'static>,
+        (eth_address, block_number_or_hash): Self::Params,
+    ) -> Result<Self::Ok, ServerError> {
+        let ts = tipset_by_block_number_or_hash(&ctx.chain_store, block_number_or_hash)?;
+        let to_address = FilecoinAddress::try_from(&eth_address)?;
+        let actor = ctx
+            .state_manager
+            .get_required_actor(&to_address, *ts.parent_state())?;
+        // Not a contract. We could try to distinguish between accounts and "native" contracts here,
```
Comment (hanabi1224) [inline src/rpc/methods/eth.rs:1305]: Comment is copied from lotus

File: src/rpc/methods/eth.rs [hunk 2]
```diff
@@ -1373,6 +1283,73 @@ impl RpcMethod<0> for EthSyncing {
...
+                            anyhow::anyhow!("GetBytecode failed: {}", api_invoc_result.error)
+                                .into(),
+                        )
+                    }
+                } else {
+                    Err(anyhow::anyhow!("no message receipt").into())
+                }
+            } else {
+                Err(anyhow::anyhow!("failed to call GetBytecode").into())
+            }
```
Comment (lemmih) [inline src/rpc/methods/eth.rs:1348]: Let-else may or may not be nicer here:
```suggestion
            let Some(api_invoc_result) = api_invoc_result else {
                return Err(anyhow::anyhow!("no message receipt").into());
            };
            let Some(msg_rct) = api_invoc_result.msg_rct else {
                return Err(anyhow::anyhow!("no message receipt").into());
            };
            if !api_invoc_result.error.is_empty() {
                return Err(
                    anyhow::anyhow!("GetBytecode failed: {}", api_invoc_result.error).into(),
                );
            }

            let get_bytecode_return: GetBytecodeReturn =
                fvm_ipld_encoding::from_slice(msg_rct.return_data().as_slice())?;
            if let Some(cid) = get_bytecode_return.0 {
                Ok(Bytes(ctx.store().get_required(&cid)?))
            } else {
                Ok(Default::default())
            }
```

**Review by lemmih — APPROVED** [review]

File: src/rpc/methods/eth/types.rs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                    let eth_addr: EthAddress = f4_addr.subaddress().try_into()?;
+                    if eth_addr.is_masked_id() {
+                        bail!(
+                            "f410f addresses cannot embed masked-ID payloads: {}",
+                            eth_addr.0
+                        );
+                    }
+                    Ok(eth_addr)
+                } else {
+                    bail!("invalid delegated address namespace in: {addr}")
```
Comment (ruseinov) [inline src/rpc/methods/eth/types.rs:60]: nit: you can get rid of this `else` block by returning early in `if let Ok...` and just bailing here.

File: src/rpc/methods/eth.rs [hunk 4]
```diff
@@ -1373,6 +1283,70 @@ impl RpcMethod<0> for EthSyncing {
...
+        (eth_address, block_number_or_hash): Self::Params,
+    ) -> Result<Self::Ok, ServerError> {
+        let ts = tipset_by_block_number_or_hash(&ctx.chain_store, block_number_or_hash)?;
+        let to_address = FilecoinAddress::try_from(&eth_address)?;
+        let actor = ctx
+            .state_manager
+            .get_required_actor(&to_address, *ts.parent_state())?;
+        // Not a contract. We could try to distinguish between accounts and "native" contracts here,
+        // but it's not worth it.
+        if !fil_actor_interface::is_evm_actor(&actor.code) {
```
Comment (ruseinov) [inline src/rpc/methods/eth.rs:1307]: nit: return early here, then you can get of the huge else block that follows.

**Review by ruseinov — APPROVED** [review]
looks good, apart from a bunch of non-blocking comments.

File: src/rpc/methods/eth.rs (hunk 4 above)
Comment (hanabi1224) [reply src/rpc/methods/eth.rs:1307]: Fixed

File: src/rpc/methods/eth/types.rs (hunk 3 above)
Comment (hanabi1224) [reply src/rpc/methods/eth/types.rs:60]: Fixed

**Review by ruseinov — APPROVED** [review]

**Pull request merged**